### ISSUE TITLE
[FireMonkey] Improvement of `GM.xmlHttpRequest` code

### ISSUE THREAD

**Issue by pintassilgo** [erosman/support#516]
Splitting latest discussion in #429 in a separated issue.

1. v2.60 introduced a bug causing `xhr.responseText` and `xhr.responseXML` to be [unavailable](https://github.com/erosman/support/issues/429#issuecomment-1285915187).
2. The issue above was fixed in v2.65, but then another bug was revealed: all the response of `GM.xmlHttpRequest` is invisibly injected to page, a severe leak that may cause high memory and CPU usage. You may notice the leak it in two ways: (1) tab icon can change to page icon of the requested URL, and (2) if the requested URL auto-plays audio, audio suddenly starts playing as reported [here](https://github.com/erosman/support/issues/429#issuecomment-1286520869) and in [this minified example](https://github.com/erosman/support/issues/429#issuecomment-1287565709).

After investigating the issue, I noticed some other things that can be improved in the code, like:
- `prepareInit()` has `url` as the first parameter for no reason, it's unused.
- `GM.xmlHttpRequest` is missing a few things for no reason, like `onreadystatechange` and `onloadstart`, available in both [Violentmonkey](https://violentmonkey.github.io/api/gm/#gm_xmlhttprequest) and [Tampermonkey](https://www.tampermonkey.net/documentation.php#GM_xmlhttpRequest).

As suggested by me in #429, `XMLHttpRequest` should run from `api.js` instead of `background.js`. Most of FM `GM.xmlHttpRequest` issues  come from Fx restrictions when passing complex objects between different contexts, like from `background` to `user_scripts api_script`. That's why current code unnecessarily needs to do things like
```js
responseText: ['', 'text'].includes(xhr.responseType) ? xhr.responseText : null
```
instead of just
```js
responseText: xhr.responseText,
```

Unlike you said there's no CORS related issue, `api.js` can perform cross-origin requests freely.

Without further ado, here are the changes I implemented in my fork and I suggest you to merge them to FireMonkey.

https://github.com/pintassilgo/forked-firefox-extensions/commit/e7b9ec45afd29573a37a578ec34f11132b0b2328

**Comment by erosman** [comment]
> prepareInit() has url as the first parameter for no reason, it's unused.

It is a leftover from earlier versions. :man_facepalming: 

```js
      // add contexual cookies, only in container/incognito
      const cookies = !init.anonymous && await browser.runtime.sendMessage({
        name,
        api: 'cookie',
        data: {url}
      });
      if (cookies) { 
        init.headers['FM-Cookie'] = (init.headers['FM-Cookie'] ? init.headers['FM-Cookie'] + '; ' : '') +
                                        cookies.map(item => `${item.name}=${item.value}`).join('; ');  
      }
```

The process was later moved elsewhere.
I will remove it.

Looking at the code, I noticed that the `async` is also a leftover from above an no longer needed.

Therefore ....

```js
// --- prepare request headers
prepareInit(init) {

.....
api.prepareInit(data.init);
.....
api.prepareInit(data);
```

There should be no need to promisfy `xmlHttpRequest`. 

In your code, there are 2 sets of `xmlHttpRequest`, one from `api.js` and from `background.js`. The issue will remain as some will be sent from api and some from bg so there is a consistency issue as the results would be different.

I have asked Mozilla engineers for further information on the CORS matter. 

---
TBH, I am waiting for Firefox to finalise MV3. Due to changes in MV3, some of the processes have to be changed as the bg will become an event page.

TM|VM also will be changing in MV3 due to the service worker in chrome.

_PS. I am working on another addon for someone, which has kept me busy. The addon will be for MV3 for both Chrome/Firefox so I have been keeping a lose eye on MV3 changes._

**Comment by pintassilgo** [comment]
> In your code, there are 2 sets of `xmlHttpRequest`, one from `api.js` and from `background.js`. The issue will remain as some will be sent from api and some from bg so there is a consistency issue as the results would be different.

There's no `XMLHttpRequest` running from bg, the only reference there is to retrieve cookies. So there's no remaining issue in this regard.

https://github.com/pintassilgo/forked-firefox-extensions/blame/e7b9ec45afd29573a37a578ec34f11132b0b2328/firemonkey/content/background.js#L970-L972

```js
      case 'xmlHttpRequest':
        await this.addCookie(e.url, e.headers, storeId);
        return e.headers;
```

I didn't touch `addCookie`, which as you know does nothing besides retrieving cookies from `browser.cookies` (unavailable for `api.js`) and storing them as `FM-Contextual-Cookie` that later will be renamed.

**Comment by pintassilgo** [comment]
> I have asked Mozilla engineers for further information on the CORS matter.

I guess you're thinking as if userScript context is the same as of api.js and it's not.

userScripts are restricted by CORS, api.js is not.

```js
// ==UserScript==
// @name             tst
// @match            http://example.com/
// @version          1.0
// ==/UserScript==

const xhr = new XMLHttpRequest();
xhr.open('GET', 'https://github.com/erosman/support/issues/516');
xhr.send();
```

The code above fails:
> Cross-Origin Request Blocked: The Same Origin Policy disallows reading the remote resource at https://github.com/erosman/support/issues/516. (Reason: CORS header ‘Access-Control-Allow-Origin’ missing).

But if you run
```js
const xhr = new XMLHttpRequest();
xhr.open('GET', 'https://github.com/erosman/support/issues/516');
xhr.send();
```
directly from api.js, you can see from browser network that it works. api.js is privileged and unrestricted for `XMLHttpRequest` just like background.js.

**Comment by erosman** [comment]
> I guess you're thinking as if userScript context is the same as of api.js and it's not.

I never thought that.
In FM, userScripts run in `userScripts` context while `apiScript` runs in `content` context (hence access to some `browser` APIs).

The issue related to `content` context which is where normal addon content pages run. 

- [Bug 1605197: Cannot send request with CORS from content scripts](https://bugzilla.mozilla.org/show_bug.cgi?id=1605197)
- [Bug 1411641: CSP 'sandbox' directive prevents content scripts from matching, due to unique origin, breaking also browser features](https://bugzilla.mozilla.org/show_bug.cgi?id=1411641)

Things may have changed since then .... that is why I wanted to get some feedback from Mozilla.

Greasemonkey runs in bg script 
- https://github.com/greasemonkey/greasemonkey/blob/efd22a93121225ada47f2fe9f021af0ab6100c21/src/bg/api-provider-source.js#L252
- https://github.com/greasemonkey/greasemonkey/blob/efd22a93121225ada47f2fe9f021af0ab6100c21/src/bg/on-user-script-xhr.js#L18

Violentmonkey code is compiled and minified so I am not sure where they run `XMLHttpRequest`.
Tampermonkey source code is not available.

**Comment by erosman** [comment]
Although nothing to do with `xmlHttpRequest`, but if you like, can you test this `api.js`?
As a result of this topic, I started to make some change to the `api.js`

[api.zip](https://github.com/erosman/support/files/9846764/api.zip)

**Comment by pintassilgo** [comment]
> It should not be async there as it blocks subsequent messages.

That's not the case. What Mozilla recommends not to do is

```js
browser.runtime.onMessage.addListener(async () => {});
```

background.js is following recommendation, I didn't change that:

https://github.com/erosman/support/blob/4416ed5f17ca71f17538e7bc3fcf0e5ae2f03a14/content/background.js#L774

`(message, sender) => this.process(message, sender)` isn't async. What is async is `API.prototype.process`. Sync functions can call async functions and that's the case. Nothing is blocked.

---

> Looking at the code, I noticed that the async is also a leftover from above an no longer needed.

> There should be no need to promisfy xmlHttpRequest.

Both fixed.

More improvements:
- `GM.xmlHttpRequest` now returns the `XMLHttpRequest` object, actually a clone just as every manager does. Also `await GM.xmlHttpRequest` was never ending.
- The object is now complete. FM v2.65 (and probably any previous version) don't allow to access/run things such as `xhr.abort()`.

diff:
https://github.com/pintassilgo/forked-firefox-extensions/commit/db88cc172a3d1cb8fd54dc6701e3ba571e381a96

**Comment by erosman** [comment]
If `responseXML` can be sent, there should be no need to parse `xhr.responseText`.

**Comment by pintassilgo** [comment]
It's not that simple. `responseXML` (`XMLDocument`) is sent but it's inaccessible (`Restricted`). There's no way to send special objects like `XMLDocument`, so it need to be created by parsing `responseText`..

Then you could ask "*Why `xhr.responseXML && `?*" Well, because the sent `xhr.responseXML` can be `null`. Without this check, `responseText` will often be unnecessarily parsed and served as `responseXML` when it shouldn't.

**Comment by erosman** [comment]
re: `responseText: xhr.responseText,` & `responseXML: xhr.responseXML...`

Have you tried a different [XMLHttpRequest.responseType](https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest/responseType) as was the case in #497?

**Comment by erosman** [comment]
> I have asked Mozilla engineers for further information on the CORS matter.

I got a reply ...

> MV2 content scripts have a bug where CORS doesn't work
> MV3 content scripts don't have that bug any more; they behave identically to the web page

**Comment by pintassilgo** [comment]
I still don't know why to trust in mere words instead of simply testing to confirm if it works or not.

If you trust *robwu*, here are his instructions on how to test:
https://bugzilla.mozilla.org/show_bug.cgi?id=1605197

> var x = new XMLHttpRequest();
> x.open("GET", "https://cors-anywhere.herokuapp.com/example.com");
> x.onload = x.onerror = e => console.log(e.type);
> x.send();

If `onload` runs, cross-site request was successful.  If `onerror` runs, then it doesn't work.

If you open console in this GitHub tab and run, of course you'll get CSP error and also `error` logged as `e.type`.

Then, to know if GM.xhr can or cannot bypass CORS restrictions, just test the following UserScript using [my fork](https://github.com/pintassilgo/forked-firefox-extensions/tree/firemonkey_2.65/firemonkey):

```js
// ==UserScript==
// @name             cors
// @match            https://github.com/erosman/support/issues/516
// @version          1.0
// ==/UserScript==

GM.xmlHttpRequest({
  url: 'https://cors-anywhere.herokuapp.com/example.com',
  onload: function (req) {
    console.log('GM.xhr: loaded');
  },
  onerror: function (req) {
    console.log('GM.xhr: error');
  },
});


var x = new XMLHttpRequest();
x.open("GET", "https://cors-anywhere.herokuapp.com/example.com");
x.onload = x.onerror = e => console.log('Normal xhr: ' + e.type);
x.send();
```

Here are the results:

> Normal xhr: error
> GM.xhr: loaded

So, as expected, xhr from api.js (my fork) can bypass CORS restrictions just like xhr from background.js (official FM). GM.xhr is privileged as it should, while normal xhr is restricted as it also should. Simple as that.

---

> Have you tried a different XMLHttpRequest.responseType as was the case in #497?

Yes, it works as expected, unlike current FireMonkey. Some examples:

- `https://imgur.com/gallery/hot/viral/page/0/hit.json` → responseType = '', responseText content (stringified JSON), responseXML `null`. The exact same result as if you run normal xhr under a `imgur.com` tab.
- `https://streamable.com/w6zts7` → responseType = '', responseText content (stringified HTML), responseXML `null`. The exact same result as if you run normal xhr under a `streamable.com` tab.
- `https://planet.kde.org/atom.xml` → responseType = '', responseText content (stringified XML), responseXML `XMLDocument`. The exact same result as if you run normal xhr under a `planet.kde.org` tab.
- `http://pudim.com.br/pudim.jpg` → responseType = '', responseText content (stringified binary), responseXML `null`. The exact same result as if you run normal xhr under a `pudim.com.br` tab.

In comparison, current FM fails in every case:

- `https://imgur.com/gallery/hot/viral/page/0/hit.json` → responseXML `DocumentFragment [ #text ]`.
- `https://streamable.com/w6zts7` → responseXML `DocumentFragment(85) [ #text, meta, #text, meta, #text, title, #text, meta, #text, meta, … ]`.
- `https://planet.kde.org/atom.xml` → responseXML `DocumentFragment [ <!-- ?xml version="1…oding="utf-8"? -->, feed ]`.
- `http://pudim.com.br/pudim.jpg` → responseXML `DocumentFragment [ #text, r� ]`.

**Comment by pintassilgo** [comment]
https://github.com/pintassilgo/forked-firefox-extensions/commit/c14ee8af7c721bcbad19622e9847f307c82af449

**Comment by erosman** [comment]
> I still don't know why to trust in mere words instead of simply testing to confirm if it works or not.
> 
> If you trust robwu, here are his instructions on how to test:
> https://bugzilla.mozilla.org/show_bug.cgi?id=1605197

I actually spoke to Rob.
I had also mentioned the bug in https://github.com/erosman/support/issues/429#issuecomment-1287635167

I think you misunderstood. 
It is working now in `content` due to a bug in mv2 (Firefox didn't want that) and it is changed in mv3. 
That was the reason I placed the XHR in bg script since it was going to be subject to CORS in content script (after the bug was fixed). However, the bug from 2019 hasn't been fixed yet.
The issue is that if we change the XHR, the behaviour will be changing soon so it wont work in future and has to be reverted back.

In fact the whole userscript API is going to be reworked under [scripting](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/scripting) API (for compatibility with Chrome).

```
https://github.com/w3c/webextensions/issues/279
https://github.com/w3c/webextensions/issues/284
```

**Comment by pintassilgo** [comment]
Bug 1605197 shows the exact opposite. CORS `XMLHttpRequest` is expected to work in content scripts (as it does on Chromium) but it doesn't. That's why a simple `XMLHttpRequest` fails in userScript, which is similar to `content`. Then scripts need to use `GM.xmlHttpRequest` instead.

But `api_script` isn't the same as `content` nor `userScript`, so it doesn't have the issue and can make cross-origin requests. That's why it's safe to do make CORS  `XMLHttpRequest` in api.js.

**Comment by pintassilgo** [comment]
You're the one misunderstanding about that. I suggest you to repeat what I just tested:

1. Install Violentmonkey in a Chromium-based browser.
2. Set `content` as default injection mode.
3. Install this demo script I suggested previously:
4. 
```js
// ==UserScript==
// @name             cors
// @match            https://github.com/erosman/support/issues/516
// @version          1.0
// @grant            GM.xmlHttpRequest
// ==/UserScript==

GM.xmlHttpRequest({
  url: 'https://cors-anywhere.herokuapp.com/example.com',
  onload: () => console.log('GM.xhr: loaded'),
  onerror: () => console.log('GM.xhr: error'),
});


var x = new XMLHttpRequest();
x.open("GET", "https://cors-anywhere.herokuapp.com/example.com");
x.onload = x.onerror = e => console.log('Normal xhr: ' + e.type);
x.send();
```

You'll realize that even `new XMLHttpRequest()` can make CORS request. you'll see `Normal xhr: load` in log.

This is what Bug 1605197 is about, the opposite of what you're thinking. Content scripts should allow CORS requests, Chromium always worked like that and Firefox doesn't because it's a bug. That's why `new XMLHttpRequest` in FM logs `Normal xhr: error`.

But **api_script isn't content nor userScript, it's different then it's not affected by the bug an can do CORS requests**.

**Comment by erosman** [comment]
> Content scripts should allow CORS requests,

Content script should (and will) get the same CORS as the webpage. 

That means CORS in `content` script will behave the same way as in `page` script.
Therefore, the XHR in `content` script will be exactly the same as the XHR in `page` script.

That is good for userscripts that want to make XHR within the domain of the page e.g. a userscript on YouTube making XHR to YouTube.
However, that could be blocked if a userscript on YouTube wants to make XHR to `example.com`.

> But api_script isn't content nor userScript, it's different. 

`apiScript` runs in `content` context. That is why is has access to all [WebExtension APIs](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/Content_scripts#webextension_apis).

> [Cross-Origin Resource Sharing (CORS)](https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS)
> 
> An example of a cross-origin request: the front-end JavaScript code served from `https://domain-a.com `uses [XMLHttpRequest](https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest) to make a request for `https://domain-b.com/data.json`.
> 
> For security reasons, browsers restrict cross-origin HTTP requests initiated from scripts. For example, `XMLHttpRequest` and the [Fetch API](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API) follow the [same-origin policy](https://developer.mozilla.org/en-US/docs/Web/Security/Same-origin_policy). This means that a web application using those APIs can only request resources from the same origin the application was loaded from unless the response from other origins includes the right CORS headers.

**Comment by erosman** [comment]
### Test script
```js
// ==UserScript==
// @name             cors
// @match            *://*.example.com/*
// @version          1.0
// @grant            GM.xmlHttpRequest
// ==/UserScript==

let url = 'https://cors-anywhere.herokuapp.com/example.com';

GM.xmlHttpRequest({
  url,
  onload: () => console.log('GM.xhr: loaded'),
  onerror: () => console.log('GM.xhr: error'),
});

url = 'https://cors-anywhere.herokuapp.com/github.com';

const xhr = new XMLHttpRequest();
xhr.open('GET', url);
xhr.onload = () => console.log('Normal xhr onload: ' + xhr);
xhr.onerror = () => console.log('Normal xhr onerror: ' + xhr);
xhr.send();
```

### Chrome Violentmonkey
- Install script
- Set `content` as default injection mode.
- Open DevTools and/or Developer Tools

![image](https://user-images.githubusercontent.com/6458876/197843750-397c5afc-592e-447a-b1b3-5324c9c07342.png)


### Firefox Violentmonkey
- Install script
- Set `content` as default injection mode.
- Open Developer Tools (`about:debugging#/runtime/this-firefox`)

![image](https://user-images.githubusercontent.com/6458876/197844937-1450bb98-2e91-4347-ae85-016e763f1225.png)


### Result
In both cases, `GM.xmlHttpRequest()` goes through and `new XMLHttpRequest()` fails.

**Comment by pintassilgo** [comment]
> > [Changes to Cross-Origin Requests in Chrome Extension Content Scripts](https://www.chromium.org/Home/chromium-security/extension-content-script-fetches/)
> > tl;dr: To improve security, cross-origin fetches will soon be disallowed from content scripts in Chrome Extensions. Such requests can be made from extension background pages instead, and relayed to content scripts when needed.

This and [bug 1578405](https://bugzilla.mozilla.org/show_bug.cgi?id=1578405) convinced me that xhr implementation should stay in bg.


> > [Cross-origin XMLHttpRequest](https://developer.chrome.com/docs/extensions/mv3/xhr/)
> > Warning
> > In Manifest V3, `XMLHttpRequest` is not supported in background pages (provided by Service Workers). Please consider using its modern replacement, `fetch()`.

But then this shows that keeping xhr in bg isn't even an option for the future.

So move it to apiScript and pray for it be kept unrestricted as it currently is OR use fetch in bg to generate fake xhr properties/methods/events entirely (don't know if it's doable).

Or even ditch GM.xhr at the cost of reduced scripts compatibility.

**Comment by erosman** [comment]
It is a dilemma. I have updated [FireMonkey Help](https://erosman.github.io/support/content/help.html#manifest-v3) with additional information about manifest v3.

TM|VM with massive user-base to a large extent set the trend in userscript implementation. They are both affected directly by changes in Chrome MV3.
- TM: Chrome 10m+ users, Firefox 500k+ users
- VM: Chrome 700k+ users, Firefox 54k+ users

Unless they decide to have separate versions for Chrome & Firefox with different features (which is not practical and breaks userscripts cross-browser-compatibility), or Chrome changes its position, the future of `GM XMLHttpRequest` is uncertain.

If the `content` scripts get the same CORS privileges as the `page` script, there is hardly any point in having `GM XMLHttpRequest` in the `content` context.

Setting certain [Forbidden header name](https://erosman.github.io/support/content/help.html#forbidden-header-name) values (e.g. Cookie, Host, Origin, Referer) would also not be possible in Chrome MV3.

There are other issues in mv3 and there will be many userscripts that will break and need to be updated accordingly.

**Comment by pintassilgo** [comment]
> Setting certain [Forbidden header name](https://erosman.github.io/support/content/help.html#forbidden-header-name) values (e.g. Cookie, Host, Origin, Referer) would also not be possible in Chrome MV3.

Even for `fetch`?

**Comment by erosman** [comment]
> Even for fetch?

Yes.... in order to set them, usersript mangers use `webRequest.onBeforeSendHeaders` with `'blocking'`.

Chrome has removed `'blocking'` option in MV3.

> [Modifying network requests](https://developer.chrome.com/docs/extensions/mv3/mv3-migration/#modifying-network-requests)
> 
> Extensions that modify network requests will need to transition from the blocking version of the [Web Request API](https://developer.chrome.com/docs/extensions/reference/webRequest/) to the new [Declarative Net Request API](https://developer.chrome.com/docs/extensions/reference/declarativeNetRequest/).

There night be a way to do something with [DNR](https://developer.chrome.com/docs/extensions/reference/declarativeNetRequest/). That remains to be seen.

However, Firefox continues to support `'blocking'` as well as XHR from bg event pages in MV3, so there wont be an issue in Firefox and FireMonkey.

**Comment by pintassilgo** [comment]
> However, Firefox continues to support `'blocking'` as well as XHR from bg event pages in MV3, so there wont be an issue in Firefox and FireMonkey.

If this is resolved, then FM should keep GM.xhr in bg. It shouldn't cripple itself due to a limitation of other browsers.

**Comment by erosman** [comment]
There will be some changes in the code in the bg event page which I will start working on once I finish the external addon project I am doing (for someone else) that runs on both Chrome & Firefox. As a result, I had t familiarise myself with all the changes in Chrome & Firefox MV3.

Currently, apiScript passes most GM functions to the bg script since that avoids unnecessary storage reads. In bg event page, that wont be possible (i.e. even in bg script it would require storage read).

I have already written the code to handle all storage functions (GM getValue, setValue, listValue, & deleteValue) in apiScript but have not applied it. The only issue is that if the page is closed while the code is trying to `setValue`, the value will not be set. This is not an issue currently with MV2 bg script.

I am waiting for the finalisation of [scripting](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/scripting) API which will be the MV3 API to inject userscripts (also available in Firefox 102+ MV2).

**Comment by pintassilgo** [comment]
I'm too lazy to redo what I did for api.js in background.js. So just the minor fix for [this](https://github.com/erosman/support/issues/429#issuecomment-1286520869), which is the same as [this](https://github.com/erosman/support/issues/429#issuecomment-1287565709).

Replace this line
https://github.com/erosman/support/blob/9e2e993af0b29288b13e5578a95cdce7f6d0914b/content/api.js#L316

by
```js
        (response.responseXML = new DOMParser().parseFromString(response.responseXML, 'application/xml'));
```

And in background.js, replace
```js
      responseXML:      ['', 'document'].includes(xhr.responseType) ? xhr.responseText : null,
```

by
```js
      responseXML:      ['', 'document'].includes(xhr.responseType) ? xhr.responseXML && xhr.responseText : null,
```
